### PR TITLE
Scrape NHIF data

### PR DESCRIPTION
**Pull request by Gathondu**
#### What does this PR do?
Add scrapers for NHIF data
Add tests for NHIF scrapers
Update README
#### Description of Task to be completed?
Add Scrapers for NHIF accredited facilities and accredited inpatient facilities
#### How should this be manually tested?
run `nosetest --nocapture`
#### Any background context you want to provide?
Initially there were no scrapers for NHIF data online and now that it has been made available, this PR includes the scrapers for the same

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,6 +7,8 @@ They retrieve data from the following sites:
 - Foreign doctors: http://medicalboard.co.ke/online-services/foreign-doctors-license-register
 - Clinical officers: http://clinicalofficerscouncil.org/online-services/retention/
 - Health Facilities: http://kmhfl.health.go.ke
+- NHIF accredited facilities: http://www.nhif.or.ke/healthinsurance/medicalFacilities
+- NHIF accredited inpatient facilities: http://www.nhif.or.ke/healthinsurance/inpatientServices
```
Comment (DavidLemayian) [inline README.md]: 💯 for the Readme update.

Comment (DavidLemayian) [reply README.md]: Thought the first link is "outpatient healthcare providers who will offer outpatient care to members of the Civil Service and Disciplined services and their declared defendants."

Comment (DavidLemayian) [reply README.md]: You can fin outpatient here - http://www.nhif.or.ke/healthinsurance/outpatientServices

File: healthtools/config.py [hunk 2]
```diff
@@ -5,7 +5,9 @@
     "DOCTORS": "http://medicalboard.co.ke/online-services/retention/?currpage={}",
     "FOREIGN_DOCTORS": "http://medicalboard.co.ke/online-services/foreign-doctors-license-register/?currpage={}",
     "CLINICAL_OFFICERS": "http://clinicalofficerscouncil.org/online-services/retention/?currpage={}",
-    "TOKEN_URL": "http://api.kmhfl.health.go.ke/o/token/"
+    "TOKEN_URL": "http://api.kmhfl.health.go.ke/o/token/",
+    "NHIF-MEDICAL-FACILITIES": "http://www.nhif.or.ke/healthinsurance/medicalFacilities",
+    "NHIF-INPATIENT": "http://www.nhif.or.ke/healthinsurance/inpatientServices"
```
Comment (DavidLemayian) [inline healthtools/config.py]: Let's change the config names to follow the standard;
```
NHIF_INPATIENT
NHIF_OUTPATIENT
NHIF_OUTPATIENT_CS
```

**Review by DavidLemayian — CHANGES_REQUESTED** [review]
Considering the feedback on inpatient, outpatient, and outpatient civil servants, you might want to rename your classes first.

File: healthtools/scrapers/nhif_inpatient_facilities.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+
+class NhifInpatientScraper(Scraper):
+    """Scraper for the NHIF accredited inpatient facilities"""
+    def __init__(self):
+        super(NhifInpatientScraper, self).__init__()
+        self.site_url = SITES["NHIF-INPATIENT"]
+        self.fields = ["hospital", "postal_addr", "beds", "branch", "category", "id"]
+        self._type = "nhif-inpatient"
+        self.s3_key = "data/nhif_accredited_inpatient_facilities.json"
```
Comment (Gathondu) [inline healthtools/scrapers/nhif_inpatient_facilities.py]: @DavidLemayian while archiving the new data to s3, it fails saying that this key doesn't exsist

File: healthtools/scrapers/nhif_inpatient_facilities.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        self.s3_historical_record_key = "data/archive/nhif_accredited_inpatient_facilities-{}.json"
```
Comment (Gathondu) [inline healthtools/scrapers/nhif_inpatient_facilities.py]: and so does this one and the rest for the nhif scrapers @DavidLemayian

File: healthtools/scrapers/nhif_inpatient_facilities.py (hunk 3 above)
Comment (DavidLemayian) [reply healthtools/scrapers/nhif_inpatient_facilities.py]: @Gathondu Did you solve this?

Let's name this file `data/nhif_inpatient.json` instead.

File: healthtools/scrapers/nhif_inpatient_facilities.py (hunk 4 above)
Comment (DavidLemayian) [reply healthtools/scrapers/nhif_inpatient_facilities.py]: @Gathondu Let's name this `data/archive/nhif_inpatient-{}.json`

File: healthtools/scrapers/nhif_outpatient_cs_facilities.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+class NhifOutpatientCsScraper(Scraper):
+    """Scraper for the NHIF outpatient facilities extended to civil servants"""
+    def __init__(self):
+        super(NhifOutpatientCsScraper, self).__init__()
+        self.site_url = SITES["NHIF-OUTPATIENT_CS"]
+        self._type = "nhif-outpatient-cs"
+        self.fields = ["code", "hospital", "nhif_branch", "job_group", "cover", "id"]
+        self.s3_key = "data/nhif_outpatient_cs_facilities.json"
+        self.s3_historical_record_key = "data/archive/nhif_outpatient_cs_facilities-{}.json"
```
Comment (DavidLemayian) [inline healthtools/scrapers/nhif_outpatient_cs_facilities.py]: Let's name these `data/nhif_outpatient_cs.json`.

File: healthtools/scrapers/nhif_outpatient_facilities.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+class NhifOutpatientScraper(NhifOutpatientCsScraper):
+    """Scraper for the NHIF accredited outpatient facilities"""
+    def __init__(self):
+        super(NhifOutpatientScraper, self).__init__()
+        self.site_url = SITES["NHIF-OUTPATIENT"]
+        self.fields = ["code", "hospital", "nhif_branch", "id"]
+        self._type = "nhif-outpatient"
+        self.s3_key = "data/nhif_accredited_outpatient_facilities.json"
+        self.s3_historical_record_key = "data/archive/nhif_accredited_outpatient_facilities-{}.json"
```
Comment (DavidLemayian) [inline healthtools/scrapers/nhif_outpatient_facilities.py]: @Gathondu Let's name these `data/nhif_outpatient.json`.

File: healthtools/tests/test_scrapers.py [hunk 7]
```diff
@@ -5,16 +5,22 @@
 from healthtools.scrapers.foreign_doctors import ForeignDoctorsScraper
 from healthtools.scrapers.clinical_officers import ClinicalOfficersScraper
 from healthtools.scrapers.health_facilities import HealthFacilitiesScraper
+from healthtools.scrapers.nhif_outpatient_cs_facilities import NhifOutpatientCsScraper
+from healthtools.scrapers.nhif_inpatient_facilities import NhifInpatientScraper
+from healthtools.scrapers.nhif_outpatient_facilities import NhifOutpatientScraper
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py]: @Gathondu Let's order these inpatient, outpatient, and outpatient-cs.

File: healthtools/tests/test_scrapers.py [hunk 8]
```diff
@@ -5,16 +5,22 @@
...
 from healthtools.config import TEST_DIR, SLACK, AWS
 
 
-class TestDoctorsScraper(unittest.TestCase):
+class TestScrapers(unittest.TestCase):
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py:14]: 💯

File: healthtools/tests/test_scrapers.py [hunk 9]
```diff
@@ -5,16 +5,22 @@
...
     def setUp(self):
         self.base_scraper = Scraper()
         self.doctors_scraper = DoctorsScraper()
         self.foreign_doctors_scraper = ForeignDoctorsScraper()
         self.clinical_officers_scraper = ClinicalOfficersScraper()
         self.health_facilities_scraper = HealthFacilitiesScraper()
+        self.nhif_accredited_scraper = NhifOutpatientCsScraper()
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py]: Let's rename to `self.nhif_outpatient_cs_scraper`.

File: healthtools/tests/test_scrapers.py [hunk 10]
```diff
@@ -59,6 +77,24 @@ def test_health_facilities_scraper_uploads_to_elasticsearch(self):
             response = self.health_facilities_scraper.upload(json.loads(data))
             self.assertEqual(response["items"][0]["index"]["_shards"]["successful"], 1)
 
+    def test_nhif_accredited_scraper_uploads_to_elasticsearch(self):
+        with open(TEST_DIR + "/dummy_files/nhif_accredited_facilities.json", "r") as my_file:
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py]: This file will need a rename.

File: scraper.py [hunk 11]
```diff
@@ -10,8 +13,14 @@
...
     clinical_officers_scraper = ClinicalOfficersScraper()
     healthfacilities_scraper = HealthFacilitiesScraper()
+    nhif_accredited_facilities = NhifOutpatientCsScraper()
+    nhif_accredited_inpatient = NhifInpatientScraper()
+    nhif_accredited_outpatient = NhifOutpatientScraper()
 
     # scraping you softly with these bots...
+    nhif_accredited_outpatient_result = nhif_accredited_outpatient.scrape_site()
+    nhif_accredited_inpatient_result = nhif_accredited_inpatient.scrape_site()
+    nhif_accredited_result = nhif_accredited_facilities.scrape_site()
```
Comment (DavidLemayian) [inline scraper.py]: Let's have this after the rest of the scrapers.

File: scraper.py [hunk 12]
```diff
@@ -10,8 +13,14 @@
     foreign_doctors_scraper = ForeignDoctorsScraper()
     clinical_officers_scraper = ClinicalOfficersScraper()
     healthfacilities_scraper = HealthFacilitiesScraper()
+    nhif_accredited_facilities = NhifOutpatientCsScraper()
+    nhif_accredited_inpatient = NhifInpatientScraper()
+    nhif_accredited_outpatient = NhifOutpatientScraper()
```
Comment (DavidLemayian) [inline scraper.py]: Let's rename these variables appropriately.

File: healthtools/tests/test_scrapers.py [hunk 13]
```diff
@@ -59,6 +77,24 @@ def test_health_facilities_scraper_uploads_to_elasticsearch(self):
...
+        with open(TEST_DIR + "/dummy_files/nhif_inpatient_facilities.json", "r") as my_file:
+            data = my_file.read()
+            response = self.nhif_inpatient_scraper.upload_data(json.loads(data))
+            self.assertEqual(response["items"][0]["index"]["_shards"]["successful"], 1)
+
+    def test_nhif_outpatient_scraper_uploads_to_elasticsearch(self):
+        with open(TEST_DIR + "/dummy_files/nhif_outpatient_facilities.json", "r") as my_file:
+            data = my_file.read()
+            response = self.nhif_outpatient_scraper.upload_data(json.loads(data))
+            self.assertEqual(response["items"][0]["index"]["_shards"]["successful"], 1)
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py]: Let's remove "facilities" in the names. It's redundant - "Inpatient", "Outpatient" already talks about a facility.

File: healthtools/config.py [hunk 14]
```diff
@@ -5,7 +5,10 @@
     "DOCTORS": "http://medicalboard.co.ke/online-services/retention/?currpage={}",
     "FOREIGN_DOCTORS": "http://medicalboard.co.ke/online-services/foreign-doctors-license-register/?currpage={}",
     "CLINICAL_OFFICERS": "http://clinicalofficerscouncil.org/online-services/retention/?currpage={}",
-    "TOKEN_URL": "http://api.kmhfl.health.go.ke/o/token/"
+    "TOKEN_URL": "http://api.kmhfl.health.go.ke/o/token/",
+    "NHIF-OUTPATIENT_CS": "http://www.nhif.or.ke/healthinsurance/medicalFacilities",
```
Comment (DavidLemayian) [inline healthtools/config.py]: Let's move this to the bottom of the list.

File: healthtools/scrapers/base_scraper.py [hunk 15]
```diff
@@ -25,6 +25,8 @@ def __init__(self):
         self.s3_key = None
         self.document_id = 0  # id for each entry, to be incremented
         self.delete_file = None  # contains docs to be deleted after scrape
+        self._type = None  # elastic search for type
+        self.NHIF_SERVICES = ["accredited", "inpatient", "outpatient"]
```
Comment (DavidLemayian) [inline healthtools/scrapers/base_scraper.py]: These should be: inpatient, outpatient, outpatient-cs

File: healthtools/scrapers/base_scraper.py [hunk 16]
```diff
@@ -292,9 +292,9 @@ def format_for_elasticsearch(self, entry):
         # all bulk data need meta data describing the data
         meta_dict = {
             "index": {
-                "_index": "index",
-                "_type": "type",
-                "_id": "id"
+                "_index": ES["index"],
```
Comment (DavidLemayian) [inline healthtools/scrapers/base_scraper.py:295]: 😱

**Review by DavidLemayian — CHANGES_REQUESTED** [review]

File: healthtools/scrapers/nhif_inpatient_facilities.py (hunk 3 above)
Comment (Gathondu) [reply healthtools/scrapers/nhif_inpatient_facilities.py]: @DavidLemayian not yet, I don't have access to the aws to create the files, on renaming, on it 👍

File: healthtools/scrapers/nhif_inpatient_facilities.py (hunk 4 above)
Comment (Gathondu) [reply healthtools/scrapers/nhif_inpatient_facilities.py]: 👍

File: healthtools/tests/test_scrapers.py (hunk 7 above)
Comment (Gathondu) [reply healthtools/tests/test_scrapers.py]: 👍

File: healthtools/tests/test_scrapers.py (hunk 8 above)
Comment (Gathondu) [reply healthtools/tests/test_scrapers.py:14]: 👍

File: healthtools/tests/test_scrapers.py (hunk 9 above)
Comment (Gathondu) [reply healthtools/tests/test_scrapers.py]: 👍

File: healthtools/tests/test_scrapers.py (hunk 10 above)
Comment (Gathondu) [reply healthtools/tests/test_scrapers.py]: 👍

File: healthtools/tests/test_scrapers.py (hunk 13 above)
Comment (Gathondu) [reply healthtools/tests/test_scrapers.py]: 👍

File: scraper.py (hunk 12 above)
Comment (Gathondu) [reply scraper.py]: 👍

File: healthtools/scrapers/base_scraper.py (hunk 16 above)
Comment (Gathondu) [reply healthtools/scrapers/base_scraper.py:295]: 😲 🏃

Comment (Gathondu) [reply healthtools/scrapers/base_scraper.py:295]: wait...it correct that one..i refactored the code

File: healthtools/tests/test_scrapers.py [hunk 17]
```diff
@@ -5,16 +5,22 @@
...
+class TestScrapers(unittest.TestCase):
     def setUp(self):
         self.base_scraper = Scraper()
         self.doctors_scraper = DoctorsScraper()
         self.foreign_doctors_scraper = ForeignDoctorsScraper()
         self.clinical_officers_scraper = ClinicalOfficersScraper()
         self.health_facilities_scraper = HealthFacilitiesScraper()
+        self.nhif_outpatient_cs_scraper = NhifOutpatientCsScraper()
+        self.nhif_inpatient_scraper = NhifInpatientScraper()
+        self.nhif_outpatient_scraper = NhifOutpatientScraper()
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py]: Let's re-order this appropriately: inpatient, outpatient, outpatient-cs.

File: healthtools/tests/test_scrapers.py [hunk 18]
```diff
@@ -35,6 +41,18 @@ def test_it_scrapes_clinical_officers_page(self):
             "http://clinicalofficerscouncil.org/online-services/retention/?currpage=1")[0]
         self.assertTrue(len(entries) == 60)
 
+    def test_it_scrapes_nhif_accredited_page(self):
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py]: This requires a rename?

File: healthtools/tests/test_scrapers.py [hunk 19]
```diff
@@ -35,6 +41,18 @@ def test_it_scrapes_clinical_officers_page(self):
...
+        self.assertTrue(len(entries) > 1)
+
+    def test_it_scrapes_nhif_inpatient_page(self):
+        entries = self.nhif_inpatient_scraper.scrape_page(1)
+        self.assertTrue(len(entries) > 1)
+
+    def test_it_scrapes_nhif_outpatient_page(self):
+        entries = self.nhif_outpatient_scraper.scrape_page(1)
+        self.assertTrue(len(entries) > 1)
+
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py:51]: Let's re-order these new functions appropriately.

File: healthtools/tests/test_scrapers.py [hunk 20]
```diff
@@ -59,6 +77,24 @@ def test_health_facilities_scraper_uploads_to_elasticsearch(self):
             response = self.health_facilities_scraper.upload(json.loads(data))
             self.assertEqual(response["items"][0]["index"]["_shards"]["successful"], 1)
 
+    def test_nhif_accredited_scraper_uploads_to_elasticsearch(self):
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py]: This requires a rename.

File: healthtools/tests/test_scrapers.py [hunk 21]
```diff
@@ -59,6 +77,24 @@ def test_health_facilities_scraper_uploads_to_elasticsearch(self):
...
+            data = my_file.read()
+            response = self.nhif_inpatient_scraper.upload_data(json.loads(data))
+            self.assertEqual(response["items"][0]["index"]["_shards"]["successful"], 1)
+
+    def test_nhif_outpatient_scraper_uploads_to_elasticsearch(self):
+        with open(TEST_DIR + "/dummy_files/nhif_outpatient.json", "r") as my_file:
+            data = my_file.read()
+            response = self.nhif_outpatient_scraper.upload_data(json.loads(data))
+            self.assertEqual(response["items"][0]["index"]["_shards"]["successful"], 1)
+
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py:91]: Let's re-order these appropriately.

File: scraper.py [hunk 22]
```diff
@@ -2,6 +2,9 @@
 from healthtools.scrapers.foreign_doctors import ForeignDoctorsScraper
 from healthtools.scrapers.clinical_officers import ClinicalOfficersScraper
 from healthtools.scrapers.health_facilities import HealthFacilitiesScraper
+from healthtools.scrapers.nhif_outpatient_cs import NhifOutpatientCsScraper
+from healthtools.scrapers.nhif_inpatient import NhifInpatientScraper
+from healthtools.scrapers.nhif_outpatient import NhifOutpatientScraper
```
Comment (DavidLemayian) [inline scraper.py]: Let's re-order these.

**Review by DavidLemayian — CHANGES_REQUESTED** [review]

**Comment by Gathondu** [comment]
@DavidLemayian the changes to the above recommendations are in order

**Pull request merged**